internal/langserver/handlers/code_action.go

```diff
@@ -0,0 +1,78 @@
+package handlers
+
+import (
+	"context"
+	"fmt"
+
+	lsp "github.com/hashicorp/terraform-ls/internal/protocol"
+)
+
+func (h *logHandler) TextDocumentCodeAction(ctx context.Context, params lsp.CodeActionParams) ([]lsp.CodeAction, error) {
+	var ca []lsp.CodeAction
+	var wanted map[lsp.CodeActionKind]bool
+
+	// TODO: get this from initialize
+	supported := map[lsp.CodeActionKind]bool{
+		`source`:                        true,
+		`source.fixAll`:                 true,
+		`source.formatAll`:              true,
+		`source.formatAll.terraform-ls`: true,
+	}
+
+	// return early if we don't support any code actions
+	if len(supported) == 0 {
+		return nil, nil
+	}
+
+	// if Only is empty, assume that the client wants all code actions
+	// else build mapping of requested and determine if supported
+	if len(params.Context.Only) == 0 {
+		wanted = supported
+	} else {
+		wanted = make(map[lsp.CodeActionKind]bool)
+		for _, only := range params.Context.Only {
+			wanted[only] = supported[only]
+		}
+	}
+
+	if len(wanted) == 0 {
+		return nil, fmt.Errorf("could not find a supported code action to execute for %s, wanted %v",
+			params.TextDocument.URI, params.Context.Only)
+	}
+
+	for action, _ := range wanted {
+		switch action {
+		case lsp.Source, lsp.SourceFixAll, `source.formatAll`, `source.formatAll.terraform-ls`:
+			c, err := terraformFmt(ctx, h, params.TextDocument.URI)
+			if err != nil {
+				continue
+			}
+			ca = append(ca, c)
+		}
+	}
+
+	return ca, nil
+}
+
+func terraformFmt(ctx context.Context, h *logHandler, uri lsp.DocumentURI) (lsp.CodeAction, error) {
+	var ca lsp.CodeAction
+
+	edits, err := formatDocument(ctx, h, uri)
+	if err != nil {
+		return ca, err
+	}
+
+	we := lsp.WorkspaceEdit{
+		Changes: map[string][]lsp.TextEdit{
+			string(uri): edits,
+		},
+	}
+
+	x := lsp.CodeAction{
+		Title:       "Format Document",
+		Kind:        lsp.SourceFixAll,
+		IsPreferred: false,
+		Edit:        we,
+	}
+	return x, nil
```

Comment:
Totally nitpick: the above blocks can be basically embedded into each other like this:

```go
	return lsp.CodeAction{
		Title:       "Format Document",
		Kind:        lsp.SourceFixAll,
		IsPreferred: false,
		Edit:        lsp.WorkspaceEdit{
			Changes: map[string][]lsp.TextEdit{
				string(uri): edits,
			},
		},
	}, nil
```